spacy/ml/models/coref.py

```diff
@@ -0,0 +1,595 @@
+                stack.extend(link for link in current_node.links if not link.visited)
+            assert len(cluster) > 1
+            clusters.append(sorted(cluster))
+    return sorted(clusters)
+
+
+def build_get_head_metadata(prefix):
+    # TODO this name is awful, fix it
+    model = Model(
+        "HeadDataProvider", attrs={"prefix": prefix}, forward=head_data_forward
+    )
+    return model
+
+
+def head_data_forward(model, docs, is_train):
+    """A layer to generate the extra data needed for the span predictor."""
+    sent_ids = []
+    head_ids = []
+    prefix = model.attrs["prefix"]
+    for doc in docs:
+        sids = model.ops.asarray2i(get_sentence_ids(doc))
+        sent_ids.append(sids)
+        heads = []
+        for key, sg in doc.spans.items():
+            if not key.startswith(prefix):
+                continue
+            for span in sg:
+                # TODO warn if spans are more than one token
+                heads.append(span[0].i)
+        heads = model.ops.asarray2i(heads)
+        head_ids.append(heads)
+    # each of these is a list with one entry per doc
+    # backprop is just a placeholder
+    # TODO it would probably be better to have a list of tuples than two lists of arrays
+    return (sent_ids, head_ids), lambda x: []
+
+
+class CorefScorer(torch.nn.Module):
+    """Combines all coref modules together to find coreferent spans.
+
+    Attributes:
+        epochs_trained (int): number of epochs the model has been trained for
+
+    Submodules (in the order of their usage in the pipeline):
+        rough_scorer (RoughScorer)
+        pw (PairwiseEncoder)
+        a_scorer (AnaphoricityScorer)
+        sp (SpanPredictor)
+    """
+
+    def __init__(
+        self,
+        dim: int,  # tok2vec size
+        dist_emb_size: int,
+        hidden_size: int,
+        n_layers: int,
+        dropout_rate: float,
+        roughk: int,
+        batch_size: int,
+    ):
+        super().__init__()
+        """
+        A newly created model is set to evaluation mode.
+
+        Args:
+            epochs_trained (int): the number of epochs finished
+                (useful for warm start)
+        """
+        self.pw = DistancePairwiseEncoder(dist_emb_size, dropout_rate)
+        # TODO clean this up
+        bert_emb = dim
+        pair_emb = bert_emb * 3 + self.pw.shape
+        self.a_scorer = AnaphoricityScorer(
+            pair_emb, hidden_size, n_layers, dropout_rate
+        )
+        self.lstm = torch.nn.LSTM(
+            input_size=bert_emb,
+            hidden_size=bert_emb,
+            batch_first=True,
+        )
+        self.dropout = torch.nn.Dropout(dropout_rate)
+        self.rough_scorer = RoughScorer(bert_emb, dropout_rate, roughk)
+        self.batch_size = batch_size
+
+    def forward(self, word_features: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        This is a massive method, but it made sense to me to not split it into
```

Comment:
We need to clean up these kind of comments because no one will know who "me" is sometime in the future ;-)